daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -2167,6 +2215,30 @@ func (devices *DeviceSet) cancelDeferredRemoval(info *devInfo) error {
 	return err
 }
 
+func (devices *DeviceSet) _cancelDeferredRemoval(info *devInfo) error {
+	logrus.Debugf("devmapper: _cancelDeferredRemoval START(%s)", info.Name())
+	defer logrus.Debugf("devmapper: _cancelDeferredRemoval END(%s)", info.Name())
+
+	var err error
+
+	// Cancel deferred remove
+	for i := 0; i < 100; i++ {
+		err = devicemapper.CancelDeferredRemove(info.Name())
+		if err != nil {
+			if err == devicemapper.ErrBusy {
+				// If we see EBUSY it may be a transient error,
+				// sleep a bit a retry a few times.
+				devices.Unlock()
```

Comment:
@cpuguy83 We have two locks. There is one global lock which is devices.Lock and then there is per device lock (devInfo.lock.Lock()). We are releasing only global lock but continuing to hold per device lock. So any other task can work on other device but not on this device. This is to make sure if we are waiting on a device, operations on other devices can continue in parallel. We are doing this all across the devmapper code and this is nothing new and should be safe.